src/js/components/Report/Report.jsx

```diff
@@ -0,0 +1,221 @@
+ *   }
+ * }
+ * ```
+ *
+ * The view definition for this is:
+ *
+ * ```
+ * function MyReport() {
+ *   return (
+ *     <Report
+ *       endpoint="/reports/my-report"
+ *       keyPrefix="reports.myReport"
+ *       columns={[
+ *         { name: 'property_name', type: 'text' }
+ *       ]}/>
+ *   )
+ * }
+ * ```
+ *
+ * The `Report` component takes care of:
+ *
+ *  1. retrieving the report data from the API at `/reports/my-report` and
+ *     passing the result as `data` to the `Table`
+ *  2. creating the `Column` collection for the `Table` based on the
+ *     localization catalog rooted at `keyPrefix`
+ *  3. implementing **single column** sorting based on simple comparison or
+ *     the optional `column.sortCallback`
+ *  4. generating the Definitions section based on the column title and
+ *     descriptions
+ *  5. dispatching `SET_CURRENT_PAGE` to update the breadcrumbs
+ *
+ * @param columns {ReportColumn} -- array of report columns in the order that
+ *  they appear in the table
+ * @param endpoint {string} -- Imbi API endpoint to retrieve the report data from
+ * @param keyPrefix {string} -- translation key prefix. This is passed to
+ *   useTranslation to retrieve the translation details for the report
+ * @param pageIcon {string} -- optional FontAwesome icon name to use on the page
+ * @param title {string} -- optional title override
+ * @param tableProps {TableProps} -- additional properties are passed to the
+ *   inner `Table` component
+ * @returns {JSX.Element}
+ * @constructor
+ */
+function Report({
+  columns,
+  endpoint,
+  keyPrefix,
+  pageIcon,
+  title,
+  ...tableProps
+}) {
+  const [state, dispatch] = useContext(Context)
+  const { t: globalT } = useTranslation()
+  const { t } = useTranslation('translation', { keyPrefix })
+
+  const [errorMessage, setErrorMessage] = useState(null)
+  const [fetched, setFetched] = useState(false)
+  const [reportData, setReportData] = useState([])
+  const [sort, setSort] = useState(['', null])
+
+  const pageTitle = title || t('title')
+
+  useEffect(() => {
+    dispatch({
+      type: 'SET_CURRENT_PAGE',
+      payload: {
+        title: pageTitle,
+        url: new URL(`/ui${endpoint}`, state.baseURL)
+      }
+    })
+  }, [])
+
+  useEffect(() => {
+    if (!fetched) {
+      httpGet(
+        state.fetch,
+        new URL(endpoint, state.baseURL),
+        ({ data }) => {
+          setReportData(data)
+          setErrorMessage(null)
+          setFetched(true)
+        },
+        ({ message }) => setErrorMessage(message)
+      )
+    }
+  }, [fetched])
+
+  useEffect(() => {
+    if (sort !== null) {
+      const [column, direction] = sort
+      setReportData(
+        [...reportData].sort((a, b) => {
+          if (a[column] === null || a[column] < b[column])
+            return direction === 'asc' ? -1 : 1
+          if (b[column] === null || b[column] < a[column])
+            return direction === 'asc' ? 1 : -1
+        })
+      )
+    }
+  }, [sort])
```

Comment:
This effect is unnecessary (see the first example [here](https://react.dev/learn/you-might-not-need-an-effect#how-to-remove-unnecessary-effects)) and causing an extra render. In the example, they recommend just sorting on each render at the top of the function. Or if we expect there to be a lot of data and we don't want to re-sort on each render, you could just remove the `sort` var as a middleman and just do the update of `reportData` right in `onSortDirection`.

Reply:
I opted for the latter ... great catch. It is frightening how much about react has leaked out of my brain. I completely forgot about the dev-tool profiler, `useMemo()` and too many other things already.

Reply:
Using a state update to update other state is one of those things that catches my eye